### ISSUE TITLE
Wrong handling of inside contour points with CompoundPath#contains() and #fillRule = 'evenodd'

### ISSUE THREAD

**Issue by lehni** [paperjs/paper.js#1116]
@iconexperience unfortunately I have not only good news, but hopefully this one will be easy to resolve:

I've improved the `CompoundPath#contains() (donut)` test to check for different combinations of `fillRule` and `resolve()`, and I'm seeing some weird behavior with the `nonzero` fill-rule.

You can reproduce this is in the test suite, by uncommenting this line of code:

https://github.com/paperjs/paper.js/blob/develop/test/tests/PathItem_Contains.js#L138

And here a simple example:

``` js
view.center = [0, 0];

var path = new CompoundPath({
    children: [
        new Path.Circle([0, 0], 50),
        new Path.Circle([0, 0], 25)
    ],
    fillColor: 'blue',
    fillRule: 'evenodd'
});

var point = new Point(-25, 0);

new Path.Circle({
    center: point,
    radius: 3,
    strokeColor: 'red'
});

console.log(path.contains(point)); // false
path.fillRule = 'nonzero';
path.reorient();
console.log(path.contains(point)); // true
```

The shape looks the same in both cases...

And here the lines of code, that handles nonzero fill rule in `contains()`:

https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.js#L269

``` js
        var winding = point.isInside(
                this.getBounds({ internal: true, handle: true }))
                    && this._getWinding(point);
        return !!(this.getFillRule() === 'evenodd' ? winding & 1 : winding);
```

Is it a problem of `getWinding()` returning the wrong winding contribution?

**Comment by iconexperience** [comment]
@lehni I am not sure if I understand this one correctly. 

When I run the test in both cases the outer circle is clockwise and the inner circle is counter-clockwise, because the automatic orientation in `insertChildren` is still doing its job.
Therefore the winding of the point is 1 in both cases and `true` is returned. I tested this against the `develop`branch. Is this the wrong code base?

**Comment by lehni** [comment]
Are you running on `develop`? There is no automatic orientation in `insertChildren` anymore, I just removed this today.

**Comment by iconexperience** [comment]
@lehni I will try again.

**Comment by lehni** [comment]
@iconexperience I just updated my comment above with more info.

**Comment by iconexperience** [comment]
@lehni I think I got it now (probably forgot to pull).
The winding before reorientation is 2, which I think is correct. Imagine we did an intersect operation, in this case the inner path would belong to the result, so it must have winding 2. 

It looks some of these tests that have a point on the contour cannot be passed with our current implementation. `getWinding()` probably needs information on what to do with points that lie on the path. In this example, the correct winding is 2 for an intersect operation, but 1 for a contains operation. The same may be necessary for the <a href="https://github.com/paperjs/paper.js/issues/1075#issuecomment-232584185">other notorious test case</a>

**Comment by lehni** [comment]
@iconexperience we already have that information in `winding.contour` I think, it just doesn't get passed this far back, but it could?

**Comment by iconexperience** [comment]
@lehni Yes, that is probably the way to go. Making `_getWinding()`return the winding object should be no problem at all, and then we can see what to do with that information.

**Comment by lehni** [comment]
I've just been trying that, but in my example above, the `contour` value is 0 for that point... Does the contour calculation need updating?

**Comment by lehni** [comment]
Hmmm.... Isn't your `onPathCount` the better measurement to know if we're on the contour or not? But I still don't understand why it the contour calculation results in 0. In that point, we clearly go from outside to inside the path, so it shouldn't be, no?

**Comment by lehni** [comment]
Ok, now I see... `windingL` is 2, and `windingR` is 1, since both circles are oriented the same way... But because it's the `evenodd` fill-rule, 1 really means 0 here. Hmmm!

**Comment by iconexperience** [comment]
Why not put the `onPathCount` into the winding object instead of `confidence`? This would be more correct, because `confidence` is an interpretation of the result, that should actually be done by the function that uses that value (in out case `propagateWinding()`.

**Comment by lehni** [comment]
Yes i was thinking the same, but doesn't that bring us straight back to this discussion? It would solve this case: https://github.com/paperjs/paper.js/issues/1075#issuecomment-232584185
But it would break that: https://github.com/paperjs/paper.js/issues/1075#issuecomment-232590361

We need to find a solution here. It comes back to bite us...

**Comment by iconexperience** [comment]
Maybe that's not really a problem if we distinguish between `contains()` and boolean operations. I think we can do this, because the boolean operations do not rely on `contains()`.

See here:
![image](https://cloud.githubusercontent.com/assets/2122773/17049666/e8350c46-4fee-11e6-8fe9-f2027affa828.png)

At the left we have a compound path, where we check if the path contains the point. It is on the path, so `contains()` should return `true`. This can be done by checking the winding AND checking if the point is on the path.

At the right we have the result of a boolean subtract operation. Here we only check the winding, which is zero, so the point is not part of the path.

Wouldn't that work?

**Comment by iconexperience** [comment]
Of course it is a bit strange if `contains()` returned `true`, but there is no stroke, so it's unclear why it returns `true` :(

**Comment by iconexperience** [comment]
I have <a href="http://sketch.paperjs.org/#S/lZFBbsMgEEWvgljZkmU5Trsh6soXiNplyILA1EaeDBbG7SLK3QupSaQuKkVII8HM//8BF07qDFzwjxGCHnjFtTNp/6U8m1QY2Bsj+GadO09uIbOPR0W5k5R6tTKmGyyaIo2kVv0OOijqEYpN01TsVtq1lE/oXrN4s+pW5adF7Bw6H7kkR9sPofcAJHn2noN3IzxmjPLjY0TSPbKzXse8iwYK4AU7ZNhjxbwydpkF21bsniii2QkXkPz6C5RWfqY2Zt3iNTqCoqyDVzSjClBsk22T797+AaQFcfcf1ctTVPEDTx7UODlLYebicLz+AA==">created a Sketch</a> for better understanding. On the left you see the compound path with stroke color set. Since we assume the contour to be part of the path, `contains()` should return true. On the right there is the same path, but without stroke color. It would be really awkward if `contains()` returned true in that case.

**Comment by iconexperience** [comment]
So here is my proposal to solve this donut issue: Add `windingL` and `windingR` to the `winding` object. In `contains()` we can check for  `evenOdd` cases if one of the windings is odd. In that case `contains()` will return `true`.

Regarding the <a href="https://github.com/paperjs/paper.js/issues/1075#issuecomment-232584185">touching circles</a>: What exactly could be the problem if we accept that the point being  outside the path is the correct result?

**Comment by lehni** [comment]
- `contains()`: I agree that contains should not return true there, and it should also not change behavior based on if there's a stroke or not. For this we have hit-tests, and they will find the stroke before the fill... I'll try your suggestion of exposing `windingL`, `wingingR` later. I like that better, because interpretation of it for `contour` can then also happen where it belongs, in the boolean code.

- touching circles: I can see a few problems with considering it outside, and I have a feeling assuming that will come to haunt us:

1. Mathematically speaking, these two circles only overlap in one point, so there is an area right above and right below this point. The point is part of the outline of both these areas, so it is to be considered inside it.

2. With the current implementation, even if i move up along either of the circles from said point by a bit (e.g. 0.001 points along the path), I still get false, and that point is even more part of the shape...

**Comment by lehni** [comment]
Hang on a minute... With `develop`, now it actually returns `true`! Could this be linked to the change of `var add = path.isClockwise() ^ dir ? 1 : -1;`?

**Comment by lehni** [comment]
Oh, wait... It's because automatic sorting isn't happening anymore now : )

**Comment by iconexperience** [comment]
I do not understand. Which automatic sorting?

**Comment by lehni** [comment]
I meant automatic reorienting of compound-path children, sorry.

**Comment by iconexperience** [comment]
Ah, I think I understand, you mean the reorientation at ìnsertChildren`. But if you reverse the orientation of the second child manually, the problem reappears, no?

**Comment by lehni** [comment]
So here [the updated sketch](http://sketch.paperjs.org/#S/nVLLasMwEPyVxRfZxBg5kItDD8E/ENpjkoMjrRNhRTKSnB6C/72SEzUPSgtd8GNndne0gy6Jak6YVMlHh44dkzxhmof83BhgPbyBwk+o9anXg+Lrxh3Ty9ZsFfhgRyG5QVXBJkIhQkMoLGphmMR7fQyGyqHxXXNKc/CvXf5aYhouBltBSekjNWb5/5TKxV9KixehmN07WiHl+yCxAoJnVJpz8sTVWmqvRfZyQBKIMVuGT3jaQTEntAKH1q21UG7lUt22Fl0G38cOnveB9LazvmiFsa4OJhcHfO1axqZfTYgGTFOfdo97z59Q64zuMG5ikJPIjndFppXVEgupDynxiWuEsreDE5jd/mZAqin1m8SidKKy66gxmvPoSUmv5CNGC0rLH+Ar5K/s3mDTTaNtUm124xc=) that shows the issue at hand, and can serve as a base for further discussion.

**Comment by lehni** [comment]
We keep moving discussion of this sub-issue from #944 and #1075 around, but let's keep it here now, so I can close #1075 that's otherwise done.

**Comment by lehni** [comment]
@iconexperience I've just implemented your proposal in https://github.com/paperjs/paper.js/issues/1116#issuecomment-234480847 with cbe41c536e889d69530deb48f555a021a8cf0711, but there are still two failing cases with the original issue here, this time with horizontally touching points.

I believe it's because we don't do the horizontal / vertical distinction in `contains()`, and how can we? We don't know that we're on the curve before running the test....

What would you suggest?

```js
view.center = [0, 0];

var path = new CompoundPath({
    children: [
        new Path.Circle([0, 0], 50),
        new Path.Circle([0, 0], 25)
    ],
    fillColor: 'blue',
    fillRule: 'evenodd'
});

var point = new Point(0, -25);

new Path.Circle({
    center: point,
    radius: 3,
    strokeColor: 'red'
});

console.log(path.contains(point)); // false
path.fillRule = 'nonzero';
path.reorient();
console.log(path.contains(point)); // true
```

**Comment by lehni** [comment]
I can confirm that this is the reason. When I pass `1` for `dir`, the other two edge cases fail. You can run the unit tests to reproduce this, I've activated the failing test now.

Is there a way to use the `isOnPath` detection to know while running `getWinding()` that we should in fact switch orientation? That would be ideal, I think.

**Comment by iconexperience** [comment]
That doesn't feel right. We will then have tangential intersections, which is quite error prone. If the ray and the curve are almost parallel, the distances between paths may appear much larger than they actually are. I am afraid that we need to find another way to solve this.

**Comment by lehni** [comment]
I think you're misunderstanding me, @iconexperience. In PathItem#contains() we don't do a tangent check and we always call `getWinding()`with dir = 0. I have some code coming up that seems to address this, curious to hear your opinion on it. Gimme 5 minutes...

**Comment by lehni** [comment]
@iconexperience see 0b11f9e2dc91d53d1a22bf4b5ad2f679d44f7d74 for my proposal. I have a few questions regarding some code simplifications, but first I'd like to hear what you think. It doesn't break anything here, and it does fix this issue.

**Comment by iconexperience** [comment]
Nice, so we could then remove the tangent check in propagate winding? Makes a lot of sense to me.

**Comment by lehni** [comment]
@iconexperience I wasn't thinking that far, but I will try that, too! Might be slower though, since this iterates over the curves until it finds the parallel tangent, then restarts with the other direction.

**Comment by lehni** [comment]
And here now my question. Currently, there is [this code](https://github.com/paperjs/paper.js/blob/develop/src/path/PathItem.Boolean.js#L414-L421):
```js
            var t = null,
                a =   po === o0 ? a0
                    : po === o3 ? a3
                    : paL > max(a0, a1, a2, a3) || paR < min(a0, a1, a2, a3)
                    ? (a0 + a3) / 2
                    : Curve.solveCubic(v, io, po, roots, 0, 1) === 1
                        ? Curve.getPoint(v, t = roots[0])[dir ? 'y' : 'x']
                        : (a0 + a3) / 2,
```
I don't fully understand this part: `(a0 + a3) / 2`
And I was wondering, what you think if we turned it into that, so `t` is always defined:
```js
            var t =   po === o0 ? 0
                    : po === o3 ? 1
                    : paL > max(a0, a1, a2, a3) || paR < min(a0, a1, a2, a3)
                    ? 0.5
                    : Curve.solveCubic(v, io, po, roots, 0, 1) === 1
                        ? roots[0]
                        : 0.5,
                a = Curve.getPoint(v, t)[dir ? 'y' : 'x'],
                winding = o0 > o3 ? 1 : -1,
                windingPrev = vPrev[io] > vPrev[io + 6] ? 1 : -1,
                a3Prev = vPrev[ia + 6];
```

**Comment by iconexperience** [comment]
The first `(a0 + a3)/2` is could also be `a0`,  `a1`, `a2`, or `a3`, just any value within the abscissa range of the curve, so I decided to take the middle of the curve. You can as well use `t=0.5`, it doesn't matter, only that using t in that case may be a tiny bit slower. The second occurance of  `(a0 + a3)/2` is just a fallback to have a value. Using `t=0.5` here should not be a problem at all.

**Comment by lehni** [comment]
@iconexperience so here's my updated code now: 27af304414154cebf32640222a133c284bc5e451

I think this is pretty clean now, and solves the first case in this issue. 

The two touching circles are still up for debate...

**Comment by lehni** [comment]
@iconexperience I've created a new branch `winding-overlaps` now, and added a first attempt of handling this: 83148a2c7278a7d8357a8ede6bbd2ab242c40e70

I don't think I understand things enough to figure it out alone thought. Is this something you could imagine to work on with me?

Here some scenarios to test this against:
```js
var cp = new CompoundPath({
    children: [
        new Path.Circle({
            center: [200, 200],
            radius: 100
        }),
        new Path.Circle({
            center: [150, 200],
            radius: 50
        })
    ],
    fillRule: 'evenodd',
    fillColor: 'blue'
});

function testPointAt(offset) {
    var point = cp.firstChild.getPointAt(offset);
    new Path.Circle({
        center: point,
        radius: 2,
        strokeColor: 'red'
    });
    console.log('contains point ' + point + ': ' + cp.contains(point));
}

testPointAt(0);
```

```js
Path.inject({ // First add chainability to Path#reverse()...
    reverse: function reverse() {
        reverse.base.call(this);
        return this;
    }
});

var cp = new CompoundPath({
    children: [
        new Path.Circle({
            center: [200, 200],
            radius: 100
        }),
        new Path.Circle({
            center: [200, 200],
            radius: 100
        }).reverse(),
        new Path.Circle({
            center: [200, 200],
            radius: 100
        }),
        new Path.Circle({
            center: [150, 200],
            radius: 50
        })
    ],
    fillRule: 'evenodd',
    fillColor: 'blue'
});

function testPointAt(offset) {
    var point = cp.firstChild.getPointAt(offset);
    new Path.Circle({
        center: point,
        radius: 2,
        strokeColor: 'red'
    });
    console.log('contains point ' + point + ': ' + cp.contains(point));
}

testPointAt(0);
```

```js
var cp = new CompoundPath({
    children: [
        new Path.Rectangle({
            point: [100, 100],
            size: [200, 200]
        }),
        new Path.Rectangle({
            point: [100, 150],
            size: [100, 100]
        })
    ],
    fillRule: 'evenodd',
    fillColor: 'blue'
});

function testPointAt(point) {
    new Path.Circle({
        center: point,
        radius: 2,
        strokeColor: 'red'
    });
    console.log('contains point ' + point + ': ' + cp.contains(point));
}

testPointAt([100, 200]);
```

```js
var cp = new CompoundPath({
    children: [
        new Path.Rectangle({
            point: [100, 100],
            size: [200, 200]
        }),
        new Path.Rectangle({
            point: [100, 100],
            size: [200, 200]
        })
    ],
    fillRule: 'nonzero',
    fillColor: 'blue'
});

function testPointAt(point) {
    new Path.Circle({
        center: point,
        radius: 2,
        strokeColor: 'red'
    });
    console.log('contains point ' + point + ': ' + cp.contains(point));
}

testPointAt([100, 200]);
```

```js
var cp = new CompoundPath({
    children: [
        new Path.Rectangle({
            point: [100, 100],
            size: [200, 200]
        }),
        new Path.Rectangle({
            point: [300, 100],
            size: [200, 200]
        })
    ],
    fillRule: 'evenodd',
    fillColor: 'blue'
});

function testPointAt(point) {
    new Path.Circle({
        center: point,
        radius: 2,
        strokeColor: 'red'
    });
    console.log('contains point ' + point + ': ' + cp.contains(point));
}

testPointAt([300, 200]);
```

**Comment by iconexperience** [comment]
@lehni 
I will go for two weeks on holiday on Saturday, but until then I hope that we will find a way to close this issue and #1109 (which has the same cause IMHO).

I will write down some thoughts that hopefully will help with cases when the point is on a path.

The colors in my examples have the following meaning:

![image](https://cloud.githubusercontent.com/assets/2122773/17590369/d5147c1e-5fd9-11e6-9481-707a417ef4b1.png)

For simplicity, the ray through the point is always cast ***horizontally***.

### Path as Boundary ###

The path can be seen as a boundary between two areas that usually have different winding numbers. In the simple example below, the point `P` is on the inner path. The area right of the point has a winding of `1`, the area left of the point has a winding of `0`.

![image](https://cloud.githubusercontent.com/assets/2122773/17590380/e1478c38-5fd9-11e6-9645-33e014b22c8f.png)

The general task is to decide, which area the point should belong to. In our current algorithm cases like these are solved by simply using the higher winding number, which in this case is `1`.

Things become difficult if

A. A point is on more than one path
B. A point is on an extremum of a path

### Point on more than one Path ###

If a point is on more than one path, the areas right and left to the point can have the same winding numbers, as in this example:

![image](https://cloud.githubusercontent.com/assets/2122773/17590672/e1e111f4-5fda-11e6-9fe3-5f6bb3e8d2f0.png)

The winding on both sides is `0`, so the winding of the point is `0`. But the point is located on the path, so is it part of the path?

### Point more than once on the same Path ###

Also, a point can be on the same path more than once, as in the following examples. In both cases thw winding of the areas right and left to the point is `0`. But at leaft in the example to the left the point should be part of the path.

![image](https://cloud.githubusercontent.com/assets/2122773/17590939/2b226664-5fdc-11e6-9818-a32710d26209.png)

### Point on Extremum of a Path ###

If the point is at an extremum of a path, thw winding on both sides can be the same, as seen below. But the point is clearly part of the path, so simply taking the maximum of the left and right winding is not sufficient.:

![image](https://cloud.githubusercontent.com/assets/2122773/17591037/8f64a9f2-5fdc-11e6-8d07-0c300aa3a972.png)

Basically the same problem arises if the extremum is not a point, but a line. In this case, the horizontal ray does not enter the area with winding `1`, so the overall winding will be determined as `0`, which is wrong if we want the path to be part of the inside.

![image](https://cloud.githubusercontent.com/assets/2122773/17591480/a1b9559c-5fde-11e6-928f-bd85afd8ca61.png)

**Comment by iconexperience** [comment]
If a point is on more than one path or more than once on the same path, we can treat the paths as if there was a tiny distance between them. 

![image](https://cloud.githubusercontent.com/assets/2122773/17592082/31aa17de-5fe1-11e6-85d7-6a25e4a877bc.png)

This way, the point would be on an area with winding `1`, hence inside of the path. The problem with this approach is that in some cases the point would be inside the path even if we do not want this:

![image](https://cloud.githubusercontent.com/assets/2122773/17593379/fa8b81ca-5fe5-11e6-8e1e-d77c0e7c24f4.png)

**Comment by iconexperience** [comment]
From the examples above it has become clear that we need to collect the windings at the point separately from the windings next to the point. We are already using this approach. The question is, what to do with the windings at the point.  More to come in two weeks (sorry).

**Comment by iconexperience** [comment]
Hiking in the Rhodope Mountains gave me plenty of time to think. And I came up with a very simple model that explains why some of the remaining edge cases fail.

### Again: Path as a Boundary ###
I mentioned above that a path usually defines a boundary between two areas of different winding numbers. This is very important, so here is another very simple example as a refresher.

![image](https://cloud.githubusercontent.com/assets/2122773/18026461/7fa42fa2-6c47-11e6-9a0d-9d5d82bb2e8c.png)

Point `P` is on directly on the path. A horizontal ray is cast to the left and right. The area to the right has a winding number of `0`, the area to the left has a winding number of `1`. The point `P` has the windings `[0, 1]`.

### Point on Paths can have two Winding Numbers ###
In the example above, the point is neither in the blue area, nor in the white area. Or if you count the edges of the areas as part of the area, the Point is in the blue area and in the white area. Either way, the point has two winding numbers, it is in an **indetermined state**.

### Which Winding to Choose? ###
If the point has more than one winding, the question arises, which winding should be chosen. In most cases this is fairly simple. Since we decided previously that a point on the contour of a filled area should be part of that area, we select the winding that will result in the point being part of that area. In the simple example above, this winding is `1`.

Our current algorithm selects the higher one of the two abolute values of the windings, which usually is correct, But I will show later that this is not always correct, but for now we can keep it like this.

### Points on Horizontal Curves ###
Casting a horizontal rays works well as long as the curve at the point is not horizontal. If the curve is horizontal, the result will get ustable.
Our algorithms already takes case of these cases, as it determines the slope at the point and casts the ray horizontally if the angle of the curve is more than 45°. If the angle is less than 45°, the ray is cast vertically. This works very well.

### The Problem with Intersections and Overlaps ###
In the example above, the point was located somewhere in the middle of a curve. If a point is on a intersection (or overlap), things can become more complicated. Here is an example:
![image](https://cloud.githubusercontent.com/assets/2122773/18026547/18ee53c4-6c4b-11e6-9d15-6ad5325d7a99.png)
As you can see, `P` is located just where four areas with different winding numbers touch. The areas have winding numbers of `1`, `0`, `1` and `2`. This means that the Point `P` has three different winding numbers `[0, 1, 2]`. And this is a very important lesson: **A point on an intersection or overlap is in an indetermined state and can have more than two different winding numbers**

**Comment by lehni** [comment]
@iconexperience welcome back! I am glad to hear that the hiking has been beneficial for you as well as paper.js :) Meanwhile, I've moved to NYC, where I am about to start teaching at Parsons. I haven't had much time to take care of paper.js for the past weeks, but will pick up work again soon.

**Comment by iconexperience** [comment]
@lehni I have not found a solution yet (this will probably require more hiking :)), but at least I understand things much better now. But maybe we can start with some small changes, thus inching forward towards a proper solution.

**Comment by iconexperience** [comment]
@lehni Maybe I have found a solution.

What we probably have to do is to distinguish between two scenarios:
1. Checking if a path contains a point (`Path.contains()`).
2. Determining the winding for curves during boolean operations.

When checking if a path contains a point, we will always return `true` if the point is on the path. Look at this example:

![image](https://cloud.githubusercontent.com/assets/2122773/18133605/921da81e-6f9b-11e6-9ea8-83f0e5636e9f.png)

<a href="http://sketch.paperjs.org/#S/jVDBisIwEP2VYU4tFOm6uCwRT/2BRY/GQ0xGGxonZZq6B/HfTV1U9uZhHsx7vPeGuSCbE6HCTUfJtlihjW7az0aghxUw/UITT30c2f2Y1BblUnM/M841rQ+umPSJn63JJsPHQMVHXVfwD8p3TfMM33m+HpaDD6GJIUo+RaOQ03jnhySxo5dyFCK+a5qf6Y0Xm6MvljiRKNj+lSzqXQVinB8HBZ8VPDtUDtqHkTRey2V+xV7IdH30nAZU2931Bg==">(here is the sketch)</a>

What is the correct return value for `p.contains(bluePoint)`? If we agree, that the correct value is `true`, then this can easily be implemented by returning `true` if the point is on the path **or** if the point has a winding number that fulfills the path's fill-rule.

In the other scenario we cannot simply treat points on the path as being part of the result path. But since we only have problems determining the correct winding at segment points and we can freely decide where we determine the winding of a propagation chain, let's simply make sure in `propagateWinding()` that we do not determine the winding at a segment point, but somewhere in the middle of a curve.

If this works (and at the moment I think it will), we will only have to slightly modify our code.

**Comment by iconexperience** [comment]
I have created a first implementation of my proposal and all tests pass!

I completely removed all the `onPathCount`, `pathWindingL`, `pathWindingR` etc. from `getWinding()` and only left the `onPath` variable. This makes `getWinding()` a bit easier.
The other change is in `PathItem._contains()` where I added a check for `winding.onPath`. If `onPath` is true, the point is always on the path, hence the path contains the point.

Unfortunately I also have problems with Git. When pushing a new local branch to GitHub, I get this error message:

*Failed with error: Unable to load Canvas module. Running in a headless context. Cannot read property 'Image' of undefinedTypeError: Cannot read property 'Image' of undefined error: failed to push some refs to 'https://github.com/iconexperience/paper.js'* 

Not sure what is going on there, but I will try another approach.

**Comment by iconexperience** [comment]
@lehni I am giving up for now. I cannot even push two simple commits to my own branch anymore without getting the error mentioned above. Git has always been difficult and frustrating to handle for me, but I am totally lost now. 

Locally I will continue development and hopyfully I can soon share what I have done, but I just do not see how I can contribute to the online project at the moment.

Sorry.

**Comment by lehni** [comment]
@iconexperience oh darn, you're struggling with the automatic unit tests (`gulp test`) that are run before pushes now. I guess I should only run them when pushing to `develop`, not any of the other branches. I need to look into it. To prevent the problem, you can always push like this:

    git push --no-verify

**Comment by sapics** [comment]
I had a same problem as @iconexperience in Windows 10.
In windows system, it is more difficult to install canvas module.
https://github.com/Automattic/node-canvas/wiki/Installation---Windows

It looks good idea to run `gulp test` only in `develop` branch!

**Comment by iconexperience** [comment]
@sapics
Yes, this is a bit too much for me. So far I could avoid it by using `gulp test:browser`, which works well. There is one failed test with Chrome (424, 99.8% identical), but I simply ignore that one.

@lehni Does this have any connections with the automatic tests failing with Node 4.5.0 in the two most recent pull requests?

**Comment by lehni** [comment]
@iconexperience 56cfdd9ab67d2a2e1a1316eee4c18c7a9cdda8c2 should solve your problems. Please let me know if works for you,.

**Comment by lehni** [comment]
Since the merging in of 3c2588fdec1035d3960b46d5c7271d6ed5e2b7ff, @iconexperience's proposed solution in https://github.com/paperjs/paper.js/issues/1116#issuecomment-243794824 is now implemented. I am going to commit some unit tests for it and finally close this issue.
